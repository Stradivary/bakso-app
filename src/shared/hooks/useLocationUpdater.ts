import { useEffect } from "react";
import { LatLng } from "leaflet";
import { useLocation } from "./useLocation";

export const UPDATE_INTERVAL = 2000;

export function useLocationUpdater({
  userId,
  userRole,
  updateLocation,
}: {
  userId: string;
  userRole: string;
  updateLocation: (userId: string, location: LatLng) => void;
}) {
  const { location } = useLocation();

  useEffect(() => {
    if (userRole !== "seller") {
      return;
    }

    const updateInterval = setInterval(() => {
      updateLocation(
        userId,
        new LatLng(location?.latitude ?? 0, location?.longitude ?? 0),
      );
    }, UPDATE_INTERVAL);

    return () => clearInterval(updateInterval);
  }, [
    userRole,
    updateLocation,
    location?.latitude,
    location?.longitude,
    userId,
  ]);

  return location;
}
