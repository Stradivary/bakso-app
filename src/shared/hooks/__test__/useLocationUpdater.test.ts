import { renderHook } from "@testing-library/react";
import { LatLng } from "leaflet";
import {
  afterAll,
  afterEach,
  beforeAll,
  beforeEach,
  describe,
  expect,
  it,
  vi,
} from "vitest";
import { UPDATE_INTERVAL, useLocationUpdater } from "../useLocationUpdater";

vi.mock("./useLocation", () => ({
  useLocation: vi
    .fn()
    .mockReturnValue({ location: { latitude: 1, longitude: 1 } }),
}));

describe("useLocationUpdater", () => {
  const mockUpdateLocation = vi.fn();
  const mockLocation = { latitude: 1, longitude: 1 };
  // Define mock coordinates type for better TypeScript support
  interface MockCoordinates {
    latitude: number | null;
    longitude: number | null;
    accuracy: number;
    altitude: null | number;
    altitudeAccuracy: null | number;
    heading: null | number;
    speed: null | number;
  }

  // Default mock coordinates
  const defaultCoords: MockCoordinates = {
    latitude: 1,
    longitude: 1,
    accuracy: 100,
    altitude: null,
    altitudeAccuracy: null,
    heading: null,
    speed: null,
  };

  // Create mock geolocation object
  const mockGeolocation = {
    getCurrentPosition: vi.fn().mockImplementation((success) =>
      Promise.resolve(
        success({
          coords: defaultCoords,
          timestamp: Date.now(),
        }),
      ),
    ),
    watchPosition: vi.fn(),
    clearWatch: vi.fn(),
  };

  // Setup before tests
  beforeAll(() => {
    Object.defineProperty(global.navigator, "geolocation", {
      value: mockGeolocation,
      writable: true,
    });
  });

  beforeEach(() => {
    vi.useFakeTimers();
  });

  // Clean up after tests
  afterAll(() => {
    vi.unstubAllGlobals();
  });

  // Optional: Reset mocks between tests
  afterEach(() => {
    vi.useRealTimers();
    vi.clearAllMocks();
  });

  it("updates location for seller role", () => {
    vi.useFakeTimers();

    renderHook(() =>
      useLocationUpdater({
        userId: "123",
        userRole: "seller",
        updateLocation: mockUpdateLocation,
      }),
    );

    vi.advanceTimersByTime(UPDATE_INTERVAL);
    expect(mockUpdateLocation).toHaveBeenCalledWith("123", expect.any(LatLng));

    vi.useRealTimers();
  });

  it("does not update location when disabled", () => {
    vi.useFakeTimers();

    renderHook(() =>
      useLocationUpdater({
        userId: "123",
        userRole: "buyer",
        updateLocation: mockUpdateLocation,
      }),
    );

    vi.advanceTimersByTime(UPDATE_INTERVAL);
    expect(mockUpdateLocation).not.toHaveBeenCalled();

    vi.useRealTimers();
  });

  it("returns location from useLocation hook", () => {
    const { result } = renderHook(() =>
      useLocationUpdater({
        userId: "123",
        userRole: "seller",
        updateLocation: mockUpdateLocation,
      }),
    );

    expect(result.current).toEqual(mockLocation);
  });
});
