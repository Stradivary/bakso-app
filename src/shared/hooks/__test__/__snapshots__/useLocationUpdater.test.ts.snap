// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useLocationUpdater > returns location from useLocation hook 1`] = `
Location {
  "assign": [Function],
  "hash": "",
  "host": "localhost:3000",
  "hostname": "localhost",
  "href": "http://localhost:3000/",
  "origin": "http://localhost:3000",
  "pathname": "/",
  "port": "3000",
  "protocol": "http:",
  "reload": [Function],
  "replace": [Function],
  "search": "",
  "toString": [Function],
}
`;

exports[`useLocationUpdater > updates location for seller role 1`] = `
{
  "rerender": [Function],
  "result": {
    "current": Location {
      "assign": [Function],
      "hash": "",
      "host": "localhost:3000",
      "hostname": "localhost",
      "href": "http://localhost:3000/",
      "origin": "http://localhost:3000",
      "pathname": "/",
      "port": "3000",
      "protocol": "http:",
      "reload": [Function],
      "replace": [Function],
      "search": "",
      "toString": [Function],
    },
  },
  "unmount": [Function],
}
`;
